src/aznfswatchdog

```diff
@@ -0,0 +1,144 @@
+    # 
+    # where the format is
+    # <blobendpoint>:/account/container LOCAL_IP blobendpoint_ip
+    # 
+    IFS=$'\n' lines=$(cat $MOUNTMAP) 
+
+    for line in $lines; do
+        if [ -z "$line" ]; then
+            continue
+        fi
+
+        l_share=$(echo "$line" | awk '{print $1}')
+        l_host=$(echo "$l_share" | cut -d: -f1)
+        l_dir=$(echo "$l_share" | cut -d: -f2)
+        l_ip=$(echo "$line" | awk '{print $2}')
+        l_nfsip=$(echo "$line" | awk '{print $3}')
+
+        if [ -z "$l_host" -o -z "$l_dir" -o -z "$l_ip" -o -z "$l_nfsip" ]; then 
+            wecho "Deleting invalid line in $MOUNTMAP: [$line]."
+            update_mountmap sed -i "/$line/d" $MOUNTMAP 
+            continue 
+        fi
+
+        # Since we added it to the MOUNTMAP file, it cannot be invalid.
+        if ! is_private_ip "$l_ip"; then
+            eecho "[FATAL] local ip ($l_ip) is invalid!"
+            continue
+        fi
+
+        # Since we added it to the MOUNTMAP file, it cannot be invalid.
+        if ! is_valid_ipv4_address "$l_nfsip" || is_private_ip "$l_nfsip"; then
+            eecho "[FATAL] Blob endpoint ip ($l_nfsip) is invalid!"
+            continue
+        fi
+
+        #
+        # Local share name. Note that we mount the local IP and not the actual
+        # Blob endpoint IP and map local IP to actual Blob endpoint IP using
+        # DNAT rule.
+        #
+        m_share="$l_ip:$l_dir"
+
+        #
+        # Delete entry from MOUNTMAP if share is unmounted.
+        # TODO: What if user unmounts and mounts before MONITOR_INTERVAL_SECS secs?
+        #
+        if ! echo "$findmnt" | grep " nfs $m_share " > /dev/null; then
+            # delete the line from MOUNTMAP file.
+            pecho "[$(date -u)] Deleting unmounted share ${m_share}."
+            update_mountmap sed -i "/ $l_ip $l_nfsip/d" $MOUNTMAP
+            delete_iptable_entry $l_ip $l_nfsip
+        fi
+
+        #
+        # Check if blob endpoint IP address changed.
+        # This is the migration check.
+        #
+        new_ip=$(resolve_ipv4 "$l_host")
+
+        # If we fail to resolve the host name, try next time.
+        if [ $? -ne 0 ]; then
+            eecho "resolve_ipv4($l_host) failed!"
+            continue
+        fi
+
+        if ! is_valid_ipv4_address "$new_ip"; then
+            eecho "resolve_ipv4($l_host) returned invalid IPv4 address ($new_ip)!"
+            continue
+        fi
+
+        #
+        # If the IP changed for the Blob endpoint, we need to update the DNAT rule.
+        # This will take care of migration/failover causing the Blob endpoint IP to change.
+        #
+        # TODO: Make this less frequent than the unmount check.
+        #
+        # echo "$l_host: new_ip=$new_ip, l_nfsip=$l_nfsip"
+        #
+        if [ "$new_ip" != "$l_nfsip" ]; then 
+            pecho "IP for $l_host changed [$l_nfsip -> $new_ip]."
+            update_mountmap sed -i "/ $l_ip $l_nfsip/d" $MOUNTMAP
```

Comment:
again
update_mountmap sed -i "/$line/d" $MOUNTMAP

in all places we can delete $line, that's safer

This also allows us to have two methods
add_mountmap
delete_mountmap

which take the entire line as argument and do the >> $MOUNTMAP thing for adding and sed -i thing for deleting. This will not have the issues with sed that you were seeing

also it's better to delete the mountmap line after other things like iptables delete etc, so that in case the program crashes/restarts after that we can correctly take care of the mountmap entry next time around

in all places